### PR TITLE
raidboss: p4s basic triggers and timeline

### PR DESCRIPTION
**Pull request by Legends0**
This includes the door boss and part two. I don't have actual logs yet, but used fflogs to generate timelines and videos for determining trigger design.

Untested.

### CODE REVIEW

**Comment by lishaobin1102** [comment]
Hi,could you tell me how i can use it although the main cactbot not release?

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+      regex: /Kothornos Kick/,
+      beforeSeconds: 5,
+      infoText: (data, _matches, output) => {
+        if (data.role === 'tank' || data.role === 'healer')
+          return;
+        return output.text!();
+      },
+      outputStrings: {
+        text: {
+          en: 'Bait Kothornos Kick?',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:81]: 'Bait jump?'

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+      },
+    },
+    {
+      id: 'P4S Kothornos Quake',
+      regex: /Kothornos Quake/,
+      beforeSeconds: 5,
+      infoText: (data, _matches, output) => output.text!(),
+      outputStrings: {
+        text: {
+          en: 'Bait Kothornos Quake?',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:92]: 'Bait Earthshakers?'

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 3]
```diff
@@ -0,0 +1,704 @@
...
+        },
+      },
+    },
+    {
+      id: 'P4S Hemitheos\'s Water IV',
+      regex: /Hemitheos's Water IV/,
+      beforeSeconds: 5,
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:99]: Is there no cast for this?

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+      id: 'P4S Northerly Shift Slash',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6A02', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '6A02', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '6A02', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '6A02', source: 'ヘスペロス', capture: false }),
+      alertText: (_data, _matches, output) => output.text!(),
+      outputStrings: {
+        text: {
+          en: 'Go North Edge',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:226]: North Cleave?

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '69D9', source: 'Hesperos' }),
+      netRegexDe: NetRegexes.startsUsing({ id: '69D9', source: 'Hesperos' }),
+      netRegexFr: NetRegexes.startsUsing({ id: '69D9', source: 'Hespéros' }),
+      netRegexJa: NetRegexes.startsUsing({ id: '69D9', source: 'ヘスペロス' }),
+      delaySeconds: (_data, matches) => parseFloat(matches.castTime) - 1,
+      infoText: (data, _matches, output) => output.text!({ role: data.actingRole }),
+      outputStrings: {
+        text: {
+          en: 'Share two ${role} orbs',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:422]: This probably doesn't need a callout.  I think this is a "you need to know this mechanic to do it" sort of thing.

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+          '69F5': 'Acid',
+          '69F6': 'Lava',
+          '69F7': 'Well',
+          '69F8': 'Levinstrike',
+        };
+        return output.text!({ pinax: pinax[matches.id] });
+      },
+      outputStrings: {
+        text: {
+          en: 'Be on ${pinax} tile',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:444]: What about "${pinax} safe" to call the most important word first?

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+      netRegex: NetRegexes.startsUsing({ id: '6A26', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '6A26', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '6A26', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '6A26', source: 'ヘスペロス', capture: false }),
+      alertText: (data, _matches, output) => data.role === 'tank' ? output.tankbustersIn!() : output.getOut!(),
+      outputStrings: {
+        tankbustersIn: {
+          en: 'Get in, tankbuster',
+        },
+        getOut: Outputs.out,
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:490]: This should say "Out (avoid tankbusters)" I think.  It's not just get out, and everybody can mitigate this.

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+      id: 'P4S Nearsight',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6A26', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '6A26', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '6A26', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '6A26', source: 'ヘスペロス', capture: false }),
+      alertText: (data, _matches, output) => data.role === 'tank' ? output.tankbustersIn!() : output.getOut!(),
+      outputStrings: {
+        tankbustersIn: {
+          en: 'Get in, tankbuster',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:488]:
```suggestion
          en: 'In (Tankbusters)',
```

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+      },
+    },
+    {
+      id: 'P4S Well Pinax',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '69D6', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '69D6', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '69D6', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '69D6', source: 'ヘスペロス', capture: false }),
+      delaySeconds: 5,
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:176]:
```
[21:28:20.488] StartsCasting 14:40019333:Hesperos:69D6:Well Pinax:40019333:Hesperos:8.70:90.00:90.00:0.00:0.00
[21:28:23.605] StartsCasting 14:40019303:Hesperos:69FE:Southerly Shift:40019303:Hesperos:7.70:99.99:99.99:0.00:-1.30
[21:28:29.477] ActionEffect 15:40019333:Hesperos:69D6:Well Pinax:E0000000::0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:::::::::::44:44:0:10000:::90.00:90.00:0.00:0.00:000163D6:0:0
[21:28:31.571] ActionEffect 15:40019303:Hesperos:69FE:Southerly Shift:40019303:Hesperos:1B:69FE8000:0:0:0:0:0:0:0:0:0:0:0:0:0:0:35230907:41780192:10000:10000:::99.99:99.99:0.00:3.14:35230907:41780192:10000:10000:::99.99:99.99:0.00:3.14:000163E2:0:1
```

There is roughly 11 seconds between Well Pinax cast start and Southerly Shift, so maybe delay 6?

Similarly, Southerly Shift is an 8s cast, so maybe delay 3?

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+      },
+    },
+    {
+      id: 'P4S Southerly Shift Knockback',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '69FE', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '69FE', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '69FE', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '69FE', source: 'ヘスペロス', capture: false }),
+      delaySeconds: 2,
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p4s.ts:333]: delaySeconds 3 here, as mentioned above?

**Review by quisquous — COMMENTED** [review]
Sorry, I have a bunch of callout wording opinions here (keeping things short, important words first, etc).  I'll run this through the emulator shortly.

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 3 above)
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts:99]: The cast is 0.7 seconds.

00:02:49.070 | Hesperos 6 begins casting  Hemitheos's Water IV on Hesperos 6 (0.70 sec) |  
00:02:50.049 | Hesperos 6 casts  Hemitheos's Water IV

File: ui/raidboss/data/06-ew/raid/p4s.ts [hunk 11]
```diff
@@ -0,0 +1,704 @@
...
+      // Strong proximity Aoe
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '69D7', source: 'Hesperos', capture: false }),
+      netRegexDe: NetRegexes.startsUsing({ id: '69D7', source: 'Hesperos', capture: false }),
+      netRegexFr: NetRegexes.startsUsing({ id: '69D7', source: 'Hespéros', capture: false }),
+      netRegexJa: NetRegexes.startsUsing({ id: '69D7', source: 'ヘスペロス', capture: false }),
+      alertText: (_data, _matches, output) => output.text!(),
+      outputStrings: {
+        text: {
+          en: 'Go to Corner',
```
[the remark this answers is not in the document]
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts:161]: Do you think the Periaktori should be changed as well from Levinstrike to Thunder?

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 4 above)
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: I agree with this. Was copied from p4n, but this makes it easier to tell knockback vs cleave.

**Comment by quisquous** [comment]
```diff
diff --git a/ui/raidboss/data/06-ew/raid/p4s.ts b/ui/raidboss/data/06-ew/raid/p4s.ts
index d87b49ce1..625cf699f 100644
--- a/ui/raidboss/data/06-ew/raid/p4s.ts
+++ b/ui/raidboss/data/06-ew/raid/p4s.ts
@@ -550,19 +550,19 @@ const triggerSet: TriggerSet<Data> = {
     },
     {
       id: 'P4S Fleeting Impulse',
-      type: 'StartsUsing',
-      netRegex: NetRegexes.startsUsing({ id: '6A1C', source: 'Hesperos' }),
-      netRegexDe: NetRegexes.startsUsing({ id: '6A1C', source: 'Hesperos' }),
-      netRegexFr: NetRegexes.startsUsing({ id: '6A1C', source: 'Hespéros' }),
-      netRegexJa: NetRegexes.startsUsing({ id: '6A1C', source: 'ヘスペロス' }),
+      type: 'Ability',
+      netRegex: NetRegexes.ability({ id: '6A1C', source: 'Hesperos' }),
+      netRegexDe: NetRegexes.ability({ id: '6A1C', source: 'Hesperos' }),
+      netRegexFr: NetRegexes.ability({ id: '6A1C', source: 'Hespéros' }),
+      netRegexJa: NetRegexes.ability({ id: '6A1C', source: 'ヘスペロス' }),
       preRun: (data, _matches) => {
-        if (typeof data.fleetingImpulseCounter === 'undefined')
-          data.fleetingImpulseCounter = 0;
-        return data.fleetingImpulseCounter++;
+        data.fleetingImpulseCounter = (data.fleetingImpulseCounter ?? 0) + 1;
       },
       infoText: (data, matches, output) => {
-        if (matches.target === data.me)
-          return output.text!({ num: data.meFleetingImpulse = data.fleetingImpulseCo
unter });
+        if (matches.target === data.me) {
+          data.meFleetingImpulse = data.fleetingImpulseCounter;
+          return output.text!({ num: data.fleetingImpulseCounter });
+        }
       },
       outputStrings: {
         text: {

```

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: Yeah, I think so too.

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 3 above)
Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p4s.ts:99]: Ok, sounds good! Just asking.  :)

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 5 above)
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: I'll shorten it to just orbs. The role positions before is probably plenty and there is a callout for their acting role relatively near.

Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: Or should the role positions callout be updated to, orb role positions?

Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: Yeah, Belone Bursts starts casting can be "Orb Role Positions" and I think this trigger can go away.

**Comment by Legends0** [comment]
Thanks for the fleeting impulse fix. I think that about wraps up until headmarker and tether information is gathered.

File: ui/raidboss/data/06-ew/raid/p4s.ts (hunk 9 above)
Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: Hmm.  Sorry, I don't think the generalizations you wrote here work for this fight.  I think it's important that knockback calls cover both knockbacks if possible.

Well Pinax is a 9 second cast (and castTime=8.7 here, as there's always 0.3 missing for whatever reason).  If you give a 5 second warning, then:
* t=0, Well Pinax starts casting
* t=3.1 Southerly Shift starts casting
* t=3.7 "Knockback" trigger 5 seconds ahead (8.7 - 5 = 3.7)
* t=5.1 (earliest a knockback call could go and still cover Southerly Shift, probably more like 5.6 is a better compromise?)
* t=5.8 "South Knockback" trigger (7.7 - 5 = 2.7, this is fine)
* t=9 Well Pinax goes off
* t=9.7 (Well Pinax knockback wears off, if used at 3.7)
* t =11.1 Southerly Shift goes off
* t=11.8 (Southerly Shift knockback wears off, if used at 5.8)

I said before delaySeconds 6 and 3, but maybe a compromise is Well Pinax 5.5 and Southerly Shift 2.5?

Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p4s.ts]: Yea, might be one where it's going to take some experience to get the timing down. I have updated it to 5.5 and 2.5.

**Comment by quisquous** [comment]
Sounds good.  I am going to give this a quick test tonight and then will merge after.

**Comment by Legends0** [comment]
I took a look at some more logs and saw that it was possible to write some triggers for the first set of Chlamys as bloodrake hits can be tracked instead of the tether.  I'm not sure if something similar can be done with the second role call later in the fight.

**Comment by quisquous** [comment]
I appreciate all the ongoing changes but I feel continuing to add changes here also makes it harder for this to converge to where I am confident to merge/release it, if that makes sense.  Each new thing on top of the old adds more risk/testing, if that helps explain my reasoning.  I think it's better to have no calls than to have "bad calls" from cactbot.

In other words, Bloodrakes look a bit complicated, and I think I would rather merge the original thought of "basic triggers and timeline" and then improve from there in followup PRs that can be tested/merged separately.

Prior to the additional changes last night, my feedback from testing this is:
* cape/sword mechanic needs to be called out immediately so you know which direction to run
* to avoid being confusing with surecast timing, I think it should say "west cape" instead of "west knockback"
* knockback timing from well is incorrect to catch both (it's too early), not sure what's wrong here
* to avoid confusion maybe the well should just say "well" and not "middle knockback" for now?

And then with more testing we can add a smarter "knockback" timed call in a followup PR that is in the right place for people to hit the button.

Does that all seem reasonable path to get this merged?

**Comment by lishaobin1102** [comment]
> I appreciate all the ongoing changes but I feel continuing to add changes here also makes it harder for this to converge to where I am confident to merge/release it, if that makes sense. Each new thing on top of the old adds more risk/testing, if that helps explain my reasoning. I think it's better to have no calls than to have "bad calls" from cactbot.
> 
> In other words, Bloodrakes look a bit complicated, and I think I would rather merge the original thought of "basic triggers and timeline" and then improve from there in followup PRs that can be tested/merged separately.
> 
> Prior to the additional changes last night, my feedback from testing this is:
> 
> * cape/sword mechanic needs to be called out immediately so you know which direction to run
> * to avoid being confusing with surecast timing, I think it should say "west cape" instead of "west knockback"
> * knockback timing from well is incorrect to catch both (it's too early), not sure what's wrong here
> * to avoid confusion maybe the well should just say "well" and not "middle knockback" for now?
> 
> And then with more testing we can add a smarter "knockback" timed call in a followup PR that is in the right place for people to hit the button.
> 
> Does that all seem reasonable path to get this merged?

Hi, if is it possible that push the timeline first? and then you can test more about callout? Thans very much

**Comment by weltall** [comment]
Giving it a spin on the last version during prog. 

Effectively bloodrakes callouts are exactly inverted. When dps has to do the mechanic it calls that tank healer has to do it and vice-versa. so yeah seems it needs more work the additional stuff as well.

**Comment by Legends0** [comment]
> I appreciate all the ongoing changes but I feel continuing to add changes here also makes it harder for this to converge to where I am confident to merge/release it, if that makes sense. Each new thing on top of the old adds more risk/testing, if that helps explain my reasoning. I think it's better to have no calls than to have "bad calls" from cactbot.
> 
> In other words, Bloodrakes look a bit complicated, and I think I would rather merge the original thought of "basic triggers and timeline" and then improve from there in followup PRs that can be tested/merged separately.
> 
> Prior to the additional changes last night, my feedback from testing this is:
> 
>     * cape/sword mechanic needs to be called out immediately so you know which direction to run
> 
>     * to avoid being confusing with surecast timing, I think it should say "west cape" instead of "west knockback"
> 
>     * knockback timing from well is incorrect to catch both (it's too early), not sure what's wrong here
> 
>     * to avoid confusion maybe the well should just say "well" and not "middle knockback" for now?
> 
> 
> And then with more testing we can add a smarter "knockback" timed call in a followup PR that is in the right place for people to hit the button.
> 
> Does that all seem reasonable path to get this merged?

I'll look more into those timings, see if I can find a log where someone did both effectively. Looks like most players are just timing for the Directional Shift knockback and getting hit by Well Pinax. Only a few that I have seen are using their immunity for Well.

The Bloodrake and Belone Coils will be better in another PR. It looks like it may be an easy fix from where it is according to @weltall, but needs more testing as I am not sure exactly what went wrong. Probably just flipped the role inserts/checks on my part.

**Comment by quisquous** [comment]
Oops, sorry, I didn't see you had updated this again and merged the timeline underneath you.

Yeah, I think in the short term to avoid confusion I'm suggesting changing the trigger wording to be infoText "well pinax" and "west cape" etc with no delay, and we'll sort out the knockback timing call later in a different PR.  It's easy to hit both visually, I just haven't figured out a way to write it as a trigger well.

We are getting hit by the well pinax if it's in the first set, but immuning it and the directional shift if it's in the second set by covering both.  (Maybe the ranged choose to get knocked back if it's spread?)

**Comment by Legends0** [comment]
I think I have the role call tethers fixed. I believe the inverse of what I was inserting needed to happen. Instead of player gets hit by first bloodrake/coils, they are debuff, it would be player gets hit by first coils, the other role gets tether and vice versa.

This should fix the calls:
```
        const role = data.party.isDPS(matches.target) ? 'dps' : 'tank/healer';

        // Second bloodrake/coils = Debuffs later
        if ((data.bloodrakeCounter ?? 0) === 2) {
          if (role === 'dps') {
            (data.debuffRole ??= []).push('healer');
            data.debuffRole.push('tank');
          } else {
            (data.debuffRole ??= []).push('dps');
          }

          // May end up needing both tether and debuff
          const debuffRole = data.debuffRole ??= [];
          if (data.tetherRole[0] === debuffRole[0])
            return output.roleEverything!({ role: roles[role] });
          return output.roleTethers!({ role: roles[role] });
        }

        // First bloodrake/coils = Tethers later
        if (role === 'dps') {
          (data.tetherRole ??= []).push('healer');
          data.tetherRole.push('tank');
        } else {
          (data.tetherRole ??= []).push('dps');
        }```

**Comment by Legends0** [comment]
> Oops, sorry, I didn't see you had updated this again and merged the timeline underneath you.
> 
No problem, I think I fixed the conflicts.

> Yeah, I think in the short term to avoid confusion I'm suggesting changing the trigger wording to be infoText "well pinax" and "west cape" etc with no delay, and we'll sort out the knockback timing call later in a different PR. It's easy to hit both visually, I just haven't figured out a way to write it as a trigger well.

I have pushed this change with 89d9453.

**Pull request merged**